### PR TITLE
META-2979 Tasks API: Limit tasks queue & keep fetching new tasks from…

### PR DESCRIPTION
**Pull request by nikhilbonte21**
https://linear.app/atlanproduct/issue/META-2979/tasks-api-limit-tasks-queue-and-keep-fetching-new-tasks-from-db

### CODE REVIEW

File: repository/src/main/java/org/apache/atlas/tasks/TaskExecutor.java [hunk 1]
```diff
@@ -40,35 +42,29 @@ public class TaskExecutor {
...
-                continue;
-            }
-
-            TASK_LOG.log(task);
+        latch = new CountDownLatch(0);
+        watcher = new TaskQueueWatcher(taskExecutorService, registry, taskTypeFactoryMap, statistics, latch);
+    }
 
-            this.executorService.submit(new TaskConsumer(task, this.registry, this.taskTypeFactoryMap, this.statistics));
+    public void addAll() {
```
Comment (mehtaanshul) [inline repository/src/main/java/org/apache/atlas/tasks/TaskExecutor.java:59]: Rename to a more relevant name

File: repository/src/main/java/org/apache/atlas/tasks/TaskExecutor.java [hunk 2]
```diff
@@ -40,35 +42,29 @@ public class TaskExecutor {
...
-    }
 
-    public void addAll(List<AtlasTask> tasks) {
-        for (AtlasTask task : tasks) {
-            if (task == null) {
-                continue;
-            }
-
-            TASK_LOG.log(task);
+        latch = new CountDownLatch(0);
```
Comment (mehtaanshul) [inline repository/src/main/java/org/apache/atlas/tasks/TaskExecutor.java:55]: Please explain the role of CountDownLatch briefly?

File: repository/src/main/java/org/apache/atlas/tasks/TaskManagement.java [hunk 3]
```diff
@@ -69,6 +70,10 @@ public TaskManagement(Configuration configuration, TaskRegistry taskRegistry) {
         createTaskTypeFactoryMap(taskTypeFactoryMap, taskFactory);
     }
 
+    public static boolean isStopped() {
```
Comment (mehtaanshul) [inline repository/src/main/java/org/apache/atlas/tasks/TaskManagement.java:73]: change method name to "hasStopped"

File: repository/src/main/java/org/apache/atlas/tasks/TaskExecutor.java (hunk 2 above)
Comment (nikhilbonte21) [reply repository/src/main/java/org/apache/atlas/tasks/TaskExecutor.java:55]: Consider it like a counter for number of tasks left in Queue
Whenever loading tasks into the Queue, it is initialized with total number of tasks that are being out into the Queue
As soon as a task is completed it is decremented.

TaskQueueWatcher holds this latch & whenever polling interval is completed it first checks if latch count is 0
if not latch.await() causes TaskQueueWatcher to wait for latch to be 0 & TaskQueueWatcher thread until then

File: repository/src/main/java/org/apache/atlas/tasks/TaskQueueWatcher.java [hunk 4]
```diff
@@ -0,0 +1,150 @@
...
+                    LOG.info("No tasks to queue, sleeping for {} ms", pollInterval);
+                }
+
+                Thread.sleep(pollInterval);
+
+            } catch (InterruptedException interruptedException) {
+                RequestContext.setWatcherThreadAlive(false);
+                LOG.error("TaskQueueWatcher: Interrupted");
+                LOG.error("TaskQueueWatcher thread is terminated, new tasks will not be loaded into the queue until next restart");
+                RequestContext.setWatcherThreadAlive(false);
```
Comment (mehtaanshul) [inline repository/src/main/java/org/apache/atlas/tasks/TaskQueueWatcher.java:95]: Duplicate

File: repository/src/main/java/org/apache/atlas/tasks/TaskQueueWatcher.java [hunk 5]
```diff
@@ -0,0 +1,152 @@
...
+                    LOG.info("No tasks to queue, sleeping for {} ms", pollInterval);
+                }
+
+                Thread.sleep(pollInterval);
+
+            } catch (InterruptedException interruptedException) {
+                LOG.error("TaskQueueWatcher: Interrupted: thread is terminated, new tasks will not be loaded into the queue until next restart");
+                break;
+            } catch (Exception e){
+                LOG.error("TaskQueueWatcher: Exception occurred");
```
Comment (nixon-atlan) [inline repository/src/main/java/org/apache/atlas/tasks/TaskQueueWatcher.java:102]: pass exception object for better tracing of exception. 
`LOG.error("TaskQueueWatcher: Exception occurred", e);`
can removed e.printStackTrace

**Review by mehtaanshul — APPROVED** [review]

**Pull request merged**